aten/src/ATen/native/mkldnn/Attention.cpp

```diff
@@ -0,0 +1,570 @@
+      partition, cp.inputLogicalTensors_, cp.outputLogicalTensors_);
+}
+
+// Execute SDPA partition
+// TODO: save output shape in cp_entry in meta kernel
+at::Tensor execute_sdpa_partition(
+    int64_t patternID,
+    std::vector<Tensor>& input_tensors,
+    cp_entry& cp,
+    bool inplace = false) {
+  int i = 0;
+
+  for (auto& each_tensor : input_tensors) {
+    cp.inputLLGATensors_[i++].set_data_handle(each_tensor.data_ptr());
+  }
+
+  at::Tensor output_tensor;
+  if (inplace) {
+    // there's no copy, so it's fine
+    output_tensor = input_tensors[0];
+  } else {
+    switch (patternID) {
+      case ONEDNN_GRAPH_SDPA_PATTERN_18_FP32:
+      case ONEDNN_GRAPH_SDPA_PATTERN_18_BF16: {
+        auto output_size = input_tensors[0].sizes().vec();
+        std::swap(output_size[1], output_size[2]);
+        std::vector<int64_t> output_strides{
+            output_size[1] * output_size[2] * output_size[3],
+            output_size[2] * output_size[3],
+            output_size[3],
+            1};
+        output_tensor = at::detail::empty_strided_cpu(
+            output_size, output_strides, input_tensors[1].scalar_type());
+        break;
+      }
+    }
+  }
+  cp.outputLLGATensors_[0].set_data_handle(output_tensor.data_ptr());
+  cp.cp_.execute(
+      onednn_graph::Stream::getStream(),
+      cp.inputLLGATensors_,
+      cp.outputLLGATensors_);
+  return output_tensor;
+}
+
+Tensor mkldnn_graph_sdpa_pattern(
+    const Tensor& query,
+    const Tensor& key,
+    const Tensor& value,
+    const c10::optional<Tensor>& scale,
+    const c10::optional<Tensor>& inverse_scale,
+    const c10::optional<Tensor>& attn_mask,
+    const c10::optional<Tensor>& causal_mask,
+    const c10::optional<Tensor>& causal_mask_value,
+    bool query_requires_transpose = false,
+    bool key_requires_transpose_twice = false,
+    bool key_requires_transpose_once = false,
+    bool value_requires_transpose = false,
+    bool output_requires_transpose_and_reorder = false) {
+  // first check cache
+  // The key has a pattern ID, as well as the shapes of input tenors
+  std::vector<int64_t> map_key;
+  map_key.reserve(1024);
+  // We use this because different thread-pools may be used
+  map_key.push_back(omp_get_max_threads());
+  // Algo ID
+  int64_t patternID = -1;
+  std::vector<Tensor> input_tensors;
+  if (output_requires_transpose_and_reorder && causal_mask.has_value() &&
+      attn_mask.has_value()) {
+    patternID = query.scalar_type() == c10::ScalarType::Float
+        ? ONEDNN_GRAPH_SDPA_PATTERN_18_FP32
+        : ONEDNN_GRAPH_SDPA_PATTERN_18_BF16;
+    input_tensors.reserve(7);
+  }
+  input_tensors.push_back(query);
+  input_tensors.push_back(key);
+  input_tensors.push_back(value);
+
+  map_key.push_back(patternID);
+
+  map_key.insert(map_key.end(), key.sizes().begin(), key.sizes().end());
+  map_key.insert(map_key.end(), query.sizes().begin(), query.sizes().end());
+  map_key.insert(map_key.end(), value.sizes().begin(), value.sizes().end());
```

Comment:
There are some redundant keys in the map since some dimensions should match. But I think it's OK for now if there's not much performance overhead.

Reply:
Good point, thanks! I'll try optimizing it.